### PR TITLE
Use cached discovery, fix climate and light device values

### PR DESCRIPTION
**Pull request by ollo69**
Hi,

I start to PR this change that basically try to limit the use of discovery method and cache retrieved data.
The intention is to share the result of one discovery call for all devices and at the same time limit the number of calls, also because in many case status device refresh fail with an error "FrequentlyInvoked".
This will also improve device detection  by HA component because cached data are used also for this.
For the moment I implemented this only for switch and cover, I think should be also extended (and can be quickly done) for light bulb, but not having a bulb I'm not sure about how to correctly update in the cache some of the state (brightness, color, etc). If you have one and time we can work also on this.

Please also consider that last pr #38 probably will break cover device because "state" attribute is not properly updated.

### CODE REVIEW

**Comment by PaulAnnekov** [comment]
I will try to take a look at this PR on the weekend. It's quite big and contains architectural changes.

**Comment by ollo69** [comment]
Thanks, I will be available for required clarifications.

File: tuyaha/devices/base.py [hunk 1]
```diff
@@ -36,15 +45,80 @@ def available(self):
...
+
+    def _control_device(self, action, param=None):
+        success, response = self.api.device_control(self.obj_id, action, param)
+        if not success:
+            self._update_data("online", False)
+        else:
+            self._last_update = datetime.now()
+        return success
+
+    def _update(self, use_discovery):
```
Comment (PaulAnnekov) [inline tuyaha/devices/base.py:63]: So previously we were using discovery only for switch, now we use it for any device type, right? And it doesn't cause backward-compatibility problems?
Why do we need this `use_discovery` then? You can safely remove this flag and QueryDevice mechanism.

File: tuyaha/devices/base.py [hunk 2]
```diff
@@ -36,15 +45,80 @@ def available(self):
     def iconurl(self):
         return self.icon
 
-    def update(self):
+    def _update_data(self, key, value, force_val=False):
+        if self.data:
+            if not force_val and self.data.get(key) is None:
```
Comment (PaulAnnekov) [inline tuyaha/devices/base.py:50]: What is this for?

File: tuyaha/devices/climate.py [hunk 3]
```diff
@@ -2,8 +2,52 @@
 
 
 class TuyaClimate(TuyaDevice):
```
Comment (PaulAnnekov) [inline tuyaha/devices/climate.py:4]: Looks like you're trying to fix another issue here, not related to Tuya server error problem. If yes, please, create a separate PR. Community will test them separately.

File: tuyaha/devices/cover.py [hunk 4]
```diff
@@ -2,27 +2,25 @@
...
         return state
 
     def open_cover(self):
         """Open the cover."""
-        success, _response = self.api.device_control(self.obj_id, "turnOnOff", {"value": "1"})
-
-        if success:
-            self.data["state"] = True
+        if self._control_device("turnOnOff", {"value": "1"}):
+            self._update_data("state", 1)
```
Comment (PaulAnnekov) [inline tuyaha/devices/cover.py:13]: `state` was `True`/`False` previously. Now it's `1`, `2`, `3`. Can you explain?

File: tuyaha/devices/light.py [hunk 5]
```diff
@@ -1,96 +1,162 @@
...
 
+"""The minimum brightness value set in the API that does not turn off the light."""
+MIN_BRIGHTNESS = 10.3
+BRIGHTNESS_STD_RANGE = (1, 255)
+
+COLTEMP_SET_RANGE = (1000, 10000)
+COLTEMP_KELV_RANGE = (2700, 6500)
+
 
 class TuyaLight(TuyaDevice):
```
Comment (PaulAnnekov) [inline tuyaha/devices/light.py:11]: It looks like you're trying to fix one more unrelated issue. If yes, please, move it to separate PR.

**Review by PaulAnnekov — CHANGES_REQUESTED** [review]

**Comment by PaulAnnekov** [comment]
@ollo69 Review is done. Waiting for fixes and comments.

File: tuyaha/tuyaapi.py [hunk 6]
```diff
@@ -1,19 +1,26 @@
...
 from requests.exceptions import HTTPError as RequestsHTTPError
+from threading import Lock
 
 from tuyaha.devices.factory import get_tuya_device
 
 TUYACLOUDURL = "https://px1.tuya{}.com"
 DEFAULTREGION = "us"
 
+MIN_DISCOVERY_INTERVAL = 60.0
+DEF_DISCOVERY_INTERVAL = 305.0
```
[the remark this answers is not in the document]
Comment (ollo69) [reply tuyaha/tuyaapi.py:16]: 305 is 5 minutes + 5 seconds.
This is because Tuya API now allow  discovery once every 300 seconds (reported as error by API itself if called with higher freq.). Because sometimes waiting exactly 300 seconds is not enought, I add 5 second for security.

File: tuyaha/devices/base.py (hunk 1 above)
Comment (ollo69) [reply tuyaha/devices/base.py:63]: This is long story. I did a lot of test, API limit currently are:
- discovery can be called once every 300 seconds
- query can be called once every 60 seconds (at all, not for device)

I put this flag to have the option to choose the call to do depending on number of device registered, and also hoping in a change Tuya side, allowind to choose the best method without having to change the library

File: tuyaha/devices/base.py (hunk 2 above)
Comment (ollo69) [reply tuyaha/devices/base.py:50]: States change are stored in device data (cache) to have states updated without having to wait discovery or query method to report it back.

File: tuyaha/devices/cover.py (hunk 4 above)
Comment (ollo69) [reply tuyaha/devices/cover.py:13]: True and False was changed in previous PR, but this is not correct. Real state is 1,2,3, indeed cover do not work anymore properly in HA after last PR

File: tuyaha/devices/climate.py (hunk 3 above)
Comment (ollo69) [reply tuyaha/devices/climate.py:4]: Ok, let me better explain how this PR was created.
This are not just some changes that I test internally but the result of the test performed in last 2 months by the HA community using the custom_component [tuya custom](https://github.com/ollo69/ha_tuya_custom) that I created.
Tuya API have a lot a limitation, I try to make it better to be able to have a better control of device as light and climate, this PR is the result of test done on this component with the wish to bring back this feature in standard component
Split to another PR  will cause to me additional because need to review the code for this device leaving only some part. more complex than leave it as is.

File: tuyaha/devices/light.py (hunk 5 above)
Comment (ollo69) [reply tuyaha/devices/light.py:11]: Same consideration already done for Climate

File: tuyaha/tuyaapi.py [hunk 7]
```diff
@@ -33,13 +40,47 @@ class TuyaSession:
 
 class TuyaApi:
 
+    def __init__(self, log_level_warn=True):
+        self._log_level_warn = log_level_warn
```
[the remark this answers is not in the document]
Comment (ollo69) [reply tuyaha/tuyaapi.py:44]: This is not true. Currently discovery (as also query) never fail if API polling interval is respected, so warning is more correct because means that you are pollind to fast. But because I'm not confidant with Tuya API, I prefer to wrap the method to be able to disable warning if somenthing change.

File: tuyaha/devices/base.py [hunk 8]
```diff
@@ -10,16 +12,23 @@ def __init__(self, data, api):
...
 
     def name(self):
         return self.obj_name
 
     def state(self):
         state = self.data.get("state")
-        if state == "true":
-            return True
-        else:
+        if state is None:
```
[the remark this answers is not in the document]
Comment (ollo69) [reply tuyaha/devices/base.py:24]: The goal was what you said, but is valid only for device with boolean state. As stated after, cover state is not boolean but numeric (1,2,3)

File: tuyaha/tuyaapi.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ollo69) [reply tuyaha/tuyaapi.py]: Anyway I review this values

**Comment by ollo69** [comment]
@PaulAnnekov,

looks like you don't have time to dedicate to this PR and to this project. If you prefer I can create a new library to use in the HA integration. Please let me know.

File: tuyaha/tuyaapi.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PaulAnnekov) [reply tuyaha/tuyaapi.py]: Just change this value to what you think will be correct and add a comment on why you decided to use this magic number.

File: tuyaha/devices/base.py (hunk 2 above)
Comment (PaulAnnekov) [reply tuyaha/devices/base.py:50]: Sorry, my question was unclear. Why do we need this particular check? Why not to always update cached value? Please, add a comment for this line.

File: tuyaha/devices/climate.py (hunk 3 above)
Comment (PaulAnnekov) [reply tuyaha/devices/climate.py:4]: Then at least add a comment with explanation of what is happening here or maybe a link to github issue. This class was very simple. Now it contains a lot of `if`s, magic numbers and calculations. We need to make it clear what's happening here for future developers.

File: tuyaha/devices/light.py (hunk 5 above)
Comment (PaulAnnekov) [reply tuyaha/devices/light.py:11]: https://github.com/PaulAnnekov/tuyaha/pull/39#discussion_r499169364

**Comment by PaulAnnekov** [comment]
I'm trying to review this repo weekly or at least once in two weeks. Please, handle my last comments, they are easy. I will be available to review them in the next 3 days.

File: tuyaha/tuyaapi.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PaulAnnekov) [reply tuyaha/tuyaapi.py]: Write something that will help next developer that will look into this code. Something that you wrote in your comment https://github.com/PaulAnnekov/tuyaha/pull/39#discussion_r491719432 but shorter, like `Tuya API limits Discovery requests to one per 300 seconds and Query requests to one per 60 seconds. We also adding +5 seconds to be sure we will not get an error`

File: tuyaha/devices/base.py (hunk 2 above)
Comment (ollo69) [reply tuyaha/devices/base.py:53]: I added a comment about this check. Anyway this it is not really important, is a flag to force a behavior that was not original implemented (write a value in device "data" that don't come from API). I put the flag just to not create a possible uncontrolled breaking change, but could also be removed.

File: tuyaha/devices/climate.py (hunk 3 above)
Comment (ollo69) [reply tuyaha/devices/climate.py:7]: Added comment as requested (as I should have already done). If still not clear i will add more of course....

File: tuyaha/devices/light.py (hunk 5 above)
Comment (ollo69) [reply tuyaha/devices/light.py:16]: As [#39 (comment)](https://github.com/PaulAnnekov/tuyaha/pull/39#discussion_r499169364)

**Comment by PaulAnnekov** [comment]
@ollo69 thanks for your work 😉

**Comment by PaulAnnekov** [comment]
PR into HA https://github.com/home-assistant/core/pull/41365
